pkg/controller/admissionchecks/multikueue/admissioncheck.go

```diff
@@ -0,0 +1,362 @@
+// Its main tasks being to:
+// - Maintain the list of remote controllers associated to each admission checks.
+// - Maintain the active state of the admission checks.
+type ACReconciler struct {
+	client          client.Client
+	helper          *multiKueueStoreHelper
+	configNamespace string
+
+	lock sync.RWMutex
+	// The list of remote controllers, indexed by the admission checks name.
+	controllers map[string]*remoteController
+	wlUpdateCh  chan event.GenericEvent
+
+	rootContext context.Context
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	updateConfigOverride func(rc *remoteController, kubeconfigs map[string][]byte) error
+}
+
+var _ reconcile.Reconciler = (*ACReconciler)(nil)
+var _ manager.Runnable = (*ACReconciler)(nil)
+
+func (a *ACReconciler) controllerFor(acName string) (*remoteController, bool) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	c, f := a.controllers[acName]
+	return c, f
+}
+
+func (a *ACReconciler) setControllerFor(acName string, c *remoteController) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if old, found := a.controllers[acName]; found {
+		old.watchCancel()
+	}
+	if c != nil {
+		a.controllers[acName] = c
+	} else {
+		delete(a.controllers, acName)
+	}
+}
+
+func (a *ACReconciler) Start(ctx context.Context) error {
+	a.rootContext = ctx
+	return nil
+}
+
+func (a *ACReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	ac := &kueue.AdmissionCheck{}
+	if err := a.client.Get(ctx, req.NamespacedName, ac); err != nil || ac.Spec.ControllerName != ControllerName {
+		if apierrors.IsNotFound(err) || !ac.DeletionTimestamp.IsZero() {
+			// stop/deleted a potential check controller
+			if cc, existing := a.controllerFor(req.Name); existing {
+				cc.watchCancel()
+				a.setControllerFor(req.Name, nil)
+				log.V(2).Info("Controller removed")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	var remCtrl *remoteController
+	inactiveReason := ""
+
+	log.V(2).Info("Reconcile AdmissionCheck")
+	if cfg, err := a.helper.ConfigFromRef(ctx, ac.Spec.Parameters); err != nil {
+		inactiveReason = fmt.Sprintf("Cannot load AC config: %s", err.Error())
+	} else {
+		kubeconfigs, err := a.getKubeConfigs(ctx, &cfg.Spec)
+		if err != nil {
+			a.setControllerFor(ac.Name, nil)
+			inactiveReason = fmt.Sprintf("Cannot load kubeconfigs: %s", err.Error())
+		} else {
+			cc, existing := a.controllerFor(ac.Name)
+			if !existing {
+				cc = newRemoteController(a.rootContext, a.client, a.wlUpdateCh)
+				a.setControllerFor(ac.Name, cc)
+			}
+
+			var err error
+			if a.updateConfigOverride != nil {
+				err = a.updateConfigOverride(cc, kubeconfigs)
+			} else {
+				err = cc.UpdateConfig(kubeconfigs)
+			}
```

Comment:
These 2 paths behave quite differently, and they even call a method on different objects.

Another alternative is to have a `clientBuilder` type or interface that accepts a `kubeconfig []byte`. This builder could be overridden in ACReconciler and passed to `newRemoteController` and from there to `remoteClient`.

This way, the unit tests for admission check are testing almost the same code as production.